pkg/volume/util.go

```diff
@@ -354,3 +354,251 @@ func UnmountViaEmptyDir(dir string, host VolumeHost, volName string, volSpec Spe
+		return make([]sets.String, 0), fmt.Errorf("key(s), operator(s) and value(s) are missing in selector.matchExpressions")
+	}
+	capacity := 0
+	for _, item := range pvc.Spec.Selector.MatchExpressions {
+		if item.Key == key && item.Operator == operator && len(item.Values) > 0 {
+			capacity++
+		}
+	}
+	if capacity == 0 {
+		return make([]sets.String, 0), fmt.Errorf("operator %q for key %q not found in selector.matchExpressions", key, operator)
+	}
+
+	ret := make([]sets.String, 0, capacity)
+	index := 0
+	for _, item := range pvc.Spec.Selector.MatchExpressions {
+		if item.Key == key && item.Operator == operator && len(item.Values) > 0 {
+			ret = append(ret, make(sets.String))
+			for _, value := range item.Values {
+				ret[index].Insert(value)
+			}
+			index++
+		}
+	}
+	return ret, nil
+}
+
+// RegionsToZones converts a set of regions to a set of zones
+// Currently cloud providers do not provide a func RegionToZone that will return all zones that are available in a given region.
+// Thats why the func RegionsToZones goes through allAvailableZones and checks if the zone is in any of the being converted regions.
+func RegionsToZones(regions, allAvailableZones sets.String, zoneToRegion func(string) (string, error)) (sets.String, error) {
+	zones := make(sets.String)
+	for candidate := range allAvailableZones {
+		if candidatesRegion, err := zoneToRegion(candidate); err != nil {
+			return make(sets.String), fmt.Errorf("failed to convert zone (%v) to a region: %v", candidate, err)
+		} else if regions.Has(candidatesRegion) {
+			zones.Insert(candidate)
+		}
+	}
+	return zones, nil
+}
+
+// PutAdminAndUserRequestsTogetherParams are input parameters for the func PutAdminAndUserRequestsTogether
+type PutAdminAndUserRequestsTogetherParams struct {
+	// PVC data structure that is used for the set of zones calculation
+	PVC *v1.PersistentVolumeClaim
+	// is the parameter zone specified in the Storage Class by an admin?
+	IsSCZoneSpecified bool
+	// in case the parameter zone is specified SCZone contains the value of this parameter, otherwise it is ""
+	SCZone string
+	// is the parameter zones specified in the Storage Class by an admin?
+	IsSCZonesSpecified bool
+	// in case the parameter zones is specified SCZones contains the value of this parameter, otherwise it is ""
+	SCZones string
+	// a func that returns a set of all available zones
+	GetAllZones func() (sets.String, error)
+	// a func that converts a zone to a region
+	ZoneToRegion func(string) (string, error)
+}
+
+// PutAdminAndUserRequestsTogether returns:
+// - either a set of zones resulting from currently available zones, allowed zone(s) by an admin in the corresponding storage class and zones preferred by the user in the selector part of the PVC
+// - or an error in case the resulting set of zones is empty or another error occurred
+func PutAdminAndUserRequestsTogether(input PutAdminAndUserRequestsTogetherParams) (sets.String, error) {
```

Comment:
It might actually be easier to make this function a class, given the complex arguments & state (e.g. gotAllAvailableZones).